### ISSUE TITLE
Assets panel instead of styles panels?

### ISSUE THREAD

**Issue by philwareham** [textpattern/textpattern#1248]
See background here: https://forum.textpattern.io/viewtopic.php?id=48528

TLDR; there was talk of changing the 'Styles' panel to a generic 'Assets' panel. Is this a planned 4.7 feature or for a later version? Bearing in mind it has direct consequences for themes it may need to be implemented now to avoid a backwards compatibility issue with any themes down the line.

If implemented now, it needs a fair amount of work to bring the language strings in line with it, and some beta testing, and some documentation. Phew!

**Comment by bloatware** [comment]
A prototype is working, serving "styles" with the correct `Content-Type` depending on their `.extension`. We could just rename "Styles" to "Assets" in 4.7 if it's ok to store all assets (`.css, .js, ...`) in the same directory. But I'm not sure everyone is fine with it, some people could prefer to keep them separated. In this case there is too much work (imo) to deliver it with 4.7. But if @Bloke or/and @NicolasGraph are takers, I will gladly help.

**Comment by Bloke** [comment]
It's working fine for me but I'd be tempted to leave the official rename for 4.7.1 or something. Can we get away with it being an undocumented feature for the time being in 4.7.0? Or do we bite this now and fix it?

The main downside is that the disk structure is `/css` which matches all our `event` things. So if we start tinkering with that, it might mean a heap of pain for the admin URLs, and our class. Basically, all the 'assets' (as were) - Pages, Forms and Styles - are handled by an 'Assets' class in 4.7.0. If we're renaming this so that 'assets' refer to all presentational content that's _not_ Forms and Pages, it's going to get confusing.

Related to that, we have the Theme deletion stance to tidy up. We have stated that anything outside of the "Theme" (Pages/Forms/Styles) are unaffected when a theme is deleted and the 'sync' checkbox is selected. That's now not the case. If you have such assets _inside_ your theme's `/css` directory they will be deleted, but if you store any such files with the same extensions outside it - even in other subdirectories - they will be left behind and (I think) a warning is issued that not everything was removed.

Is this desirable? Should we just update the pophelp to reflect this? Or is it something we need to handle better from the outset?

Alternative: revert this excellent work and sideline it all to a branch, then introduce it in 4.7.1 as a "feature fix"?

**Comment by Bloke** [comment]
One thing I have discussed with Nicolas, btw, is that if we're going to be moving our admin panels over to a better MVC-ish structure with new widgets and whatnot in future versions, the "Asset" handling (Pages/Forms/Styles) should ideally _not_ be in the `vendors/Skin` directory. Then we can use the same Page Form and Css (ahem, Asset, whatever) classes for their respective admin panels too. If we keep them where they are now and people start using them in plugins it's going to cause confusion if we move them next version.

But if we're renaming anything anyway, it probably won't matter.

**Comment by NicolasGraph** [comment]
I don't have so much time these days but here are my thoughts.

> […] But I'm not sure everyone is fine with it, some people could prefer to keep them separated.

I would try to keep things tidy with an _Assets_ tab/directory and some asset types related groups/subdirectories as for form types.

> […] We have stated that anything outside of the "Theme" (Pages/Forms/Styles) are unaffected when a theme is deleted and the 'sync' checkbox is selected. That's now not the case. […]
Should we just update the pophelp to reflect this? […]

Yes, I think it would be enough.

Now, I guess a small problem is about vocabulary… I use Asset in the Skin package as a generic term defining pages or forms or styles, etc. I guess we should clarify this if we introduce an Assets tab/folder.
Finally, Would the "css" related things in the core ($event, etc.) and the DB be updated? Could we have the css/asset events working together for some time?

**Comment by Bloke** [comment]
> I guess a small problem is about vocabulary… I use Asset in the Skin package as a generic term defining pages or forms or styles, etc. I guess we should clarify this if we introduce an Assets tab/folder.

Yes, that's what I was alluding to above. It could get mighty confusing if we publicly referred to anything _not_ Pages and Forms as "Assets" but the code used that termionology for all parts of the theme.

Once we decide on an approach, it may be nice if we could have the dual names working side-by-side for a version or two. Not sure how feasible that would be, given that Txp's event system and menu bar only uses one `$event`. Haven't looked into it.

**Comment by bloatware** [comment]
Not sure to understand everything, but if "Assets" class name is taken, we can call the new one "Resources" or whatever. Only the new tag might be called `<txp:assets />`, or we could even enhance `<txp:output_form />`, who knows. But I feel that storing all resources (CSS, JS etc) in the same directory can be confusing, code and usability wise. I would leave it as experimental feature (under `Styles` tab/directory) atm and develop `Resources` later in parallel.

**Comment by philwareham** [comment]
FWIW I think the move to a general assets tab should be done at 4.8 *as long as the current structure of themes won't get broken by it.* We are way too close to a release to retool everything again.

i.e. as long as `pages`, `forms`, `styles` can still be allowed if an `resources` (or whatever it'll be called) directory is introduced in future then it's fine.

On a separate note - I'd rather something else than `assets` was used as the name of this future feature. `assets` is too widely used for all manner of items in modern web design and I wouldn't want Textpattern to try and read my `assets` folder. It'd need to be something more unique/specific.

**Comment by Bloke** [comment]
> as long as the current structure of themes won't get broken by it

That's the million dollar statement. The only way I can see to move forward is:

1. if your _resources_ (or whatever we end up calling them) end up populating a directory per type. `/css` and `/js` and `/txt` and `/svg` etc. That'd still be slightly odd transitioning from 4.7 (all resources in the `/css` dir) -> 4.8 (resources in separate subdirs).
2. we store them all together and move all files to a different subdirectory (e.g. `/resources`).

Either way, we would have to probably permit them to appear in either place for a time, and that will require exceptions in the code because of the tight coupling between admin-side and `/themes` directory.

There might be other ways to handle it, but we need to be sure that there's at least some sort of workable migration plan post 4.7.0, otherwise I'd favour sidelining it (as awesome as it is) into a separate branch for now and thinking a bit more about it for a 4.7.1+ release.

**Comment by bloatware** [comment]
I think we'll need a migration plan either way, be it for 4.7.0 or 4.7.1. Currently, "css" directory is called `styles`, so, if we take the way 1, it should be renamed to `css` (we could make an exception, though). The way 2 yields renaming it to something assetish and possibly creating sub-directories.

If people have already started using 4.7 in production, we'd better postpone it to 4.7.1+. Otherwise, the way 2 could be as easy as replacing `$dir = 'styles'` with `$dir = 'resources'` in `Css.php` (I leave @Bloke and @NicolasGraph to confirm). Anyway, sidelining is very easy, no need for a branch: just comment out `$mimeTypes` array content in `Css.php`. Those who want to test it can modify this content (and then caveat utilitor). Or we could create an Easter egg preference.

**Comment by Bloke** [comment]
Agreed. Migration plan required. Whatever everyone thinks is the best/least disruptive.

I think you're right that just changing the name in the base class is sufficient to trigger the new name if everything is being stored in one dir.

I'm fine with a hidden pref. With this many Easter Eggs in 4.7.0, we ought to subtitle this release "chocoholic".

**Comment by NicolasGraph** [comment]
> Otherwise, the way 2 could be as easy as replacing $dir = 'styles' with $dir = 'resources' in Css.php

> I think you're right that just changing the name in the base class is sufficient to trigger the new name if everything is being stored in one dir.

Indeed, it should work.

**Comment by bloatware** [comment]
I was not able to find a reliable way to initialize `Css` class static variables from a pref, `Skin` classes hierarchy is too complex for my understanding. So -1 egg, +1 for postponing to 4.7.1+, unless someone shows me the light.

**Comment by NicolasGraph** [comment]
I'm gonna take a look…

**Comment by NicolasGraph** [comment]
Which static variables would you like to initialize and from which kind of value?

**Comment by bloatware** [comment]
@NicolasGraph thanks! Take a look at `Css.php`, I need to populate `$mimeTypes` from some pref. FYI, I have tried to init it in autoloader, adding
```
        public static function _init() {
            static::$mimeTypes = parse_ini_string(implode(n, do_list_unique(get_pref('assets_mimetypes'))));
        }
```
to `Css` class and calling `_init` method in Loader::load(). This works on the public side, but seems to fail on theme export/import.

**Comment by NicolasGraph** [comment]
I'd like to test some code but I'm not sure how…
With the `$Mimetypes` property filled, on the public side, adding a 'js' extension to a style should modify its header when outputted with the css tag, and on the admin side js files should be imported/exported normally, right?

**Comment by bloatware** [comment]
Few `<txp:css />` details [here](https://forum.textpattern.io/viewtopic.php?pid=311054#p311054).

**Comment by NicolasGraph** [comment]
Thanks @bloatware, I didn't remember about the `type` attribute and tried `extension`.
For now I have the same problem as you: my code does work on the public side but not on the admin side. I'll try do find why this evening, or tomorrow morning.

**Comment by NicolasGraph** [comment]
Can anyone remember me how to disable the async mode or to see PHP errors on the Skin tab using this mode please?

**Comment by NicolasGraph** [comment]
The commit above should work as expected but I would prefer that @bloatware or @Bloke confirm it. I can open a PR to whatever branch you'd like if needed.

However, I would just like to make you notice that as for now the way mime types work break some features for files named like `name.type.css`. It would be probably stupid to do that but if you have a `test.js.css` file, it would be imported as a stylesheet named `test.js`, and because of the `.js` extension, on the next export, it would be exported as `test.js`, not `test.js.css`. And `test.js.css` would not even be deleted on sync.

I guess that a way to avoid that – part of at least – would be to introduce a dedicated Resource class to let the Css one play well with all `.css` files. With a few more work more, I think I could make a new class work in parallel with the Css class, using the same dir – or another one ?

**Comment by bloatware** [comment]
> @bloatware, you can set the $mimetypes property from your pref here.

Thanks @NicolasGraph, that should work. But I still would like to understand why initializing `$mimeTypes` on class autoload does not suffice.

The idea was to release this useful functionality as experimental and opt-in. Those who wish could create a pref (e.g. via a plugin) and test it, knowing that it could evaluate later. Anyway, we will need a migration plan to integrate the present `css` into future `resource`. But, personally, I'm leaning towards postponing it to 4.7.1+, there are still some points I don't understand.

**Comment by NicolasGraph** [comment]
@bloatware, could you show how you wanted to call the _init method in the loader please?

**Comment by Bloke** [comment]
Is it due to the way the class is instantiated via our autoloader? If we're using static stuff isn't it better to implement `ReusableInterface`? Maybe that's why it works if it's checked directly. I'm a bit wary of instantiating directly without going through `Txp::get('\Textpattern\Class')` first, though.

That said, the class ought to be able to handle static class variables inside it just fine, unless it's being reinitialized somewhere along the line.

**Comment by bloatware** [comment]
Sure, I've added this inside `if_class_exists` of `load()` method in `Loader.php`:
```
if_method_exists($request, '_init') {
    $request::_init(); // plus trace
}
```
Then I've added  a static `_init()` method with `$mimeTypes` initialization to `Css`. I see a trace of it on both public and admin sides, but only the public side works properly.

**Comment by NicolasGraph** [comment]
> Thanks @NicolasGraph, that should work. But I still would like to understand why initializing $mimeTypes on class autoload does not suffice.

I can't see the Css class loaded in the skin tab… I'm not familiar with the loader but I guess that is why you can't make it work.

**Comment by NicolasGraph** [comment]
@bloatware, can you confirm me that the above code works as expected?

**Comment by Bloke** [comment]
> It does if the variable is hard-coded, but we would like to populate it from a pref.

Yes, which is only an issue because it's static... right?

Can we take a step back and try to find a way to do it in a conventional non-static fashion? I'm still a trifle concerned that we just make everything static that we want to share or cache. And, yes, I know PHP (and/or our codebase) isn't really geared up for doing it any other way, but over time I'd like to try and break this cycle and introduce something better, reserving statics for only when we truly need them.

I guess it's semantics, but some kind of cache manager class would encapsulate a lot of this in future versions. So if we need to store something statically, use the singleton whose job is designed to handle all of this stuff.

That way, we have one blob in memory whose purpose is to arbitrate (get/set) static content inside its own registry. Far easier to keep track of than fifty objects that each have one or two static variables in them, and probably easier to debug too, because you can inspect the state of the cache object and minimise multi-thread/multi-user race conditions via a queue.

Dunno, maybe that won't work. Just thinking out loud.

**Comment by bloatware** [comment]
> @bloatware, can you confirm me that the above code works as expected?

Thanks @NicolasGraph, it seems to work in my tests. I guess calling a class function triggers the loader, unlike a static variable. So we have a solution in this case, but it would be even better if we could ensure somehow that all classes are initialized before their use. Thanks again, ace work!

**Comment by NicolasGraph** [comment]
> Thanks @NicolasGraph, it seems to work in my tests.

Do you need me to push the code anywhere?

> I guess calling a class function triggers the loader, unlike a static variable.

Yes, that's also my guess.

> So we have a solution in this case, but it would be even better if we could ensure somehow that all classes are initialized before their use.

I'm not sure to understand what you think about.

**Comment by Bloke** [comment]
Nice sleuthing! Can't see any drawbacks from an operational standpoint. As long as a class is returned as an instance, it shouldn't matter if it's initialized with options or not. Perhaps the only reason for the differentiator was that the Reflection interface has baggage - e.g. weight/performance implications?

**Comment by bloatware** [comment]
> Perhaps the only reason for the differentiator was that the Reflection interface has baggage - e.g. weight/performance implications?

Jukka only knows :-) Funny that **new** does not seem to call the constructor immediately: when does it happen then?

**Comment by Bloke** [comment]
> Funny that new does not seem to call the constructor immediately: when does it happen then?

The PHP development only knows :-)

**Comment by bloatware** [comment]
Well, should we introduce a new interface that will call the constructor whatever options? Then `Css` would implement this and `ReusableInterface` perhaps?

**Comment by Bloke** [comment]
That would make sense, yeah. And less intrusive than doing it all the time, just in case there are some hidden performance or deep-copy issues with using the Reflection interface.

I really need to wrap my head round all this adapter/container code one day.

**Comment by bloatware** [comment]
It's one of these moments... I have registered the `$mimeType` injecting plugin to a wrong event, which has biased all testing. Sorry! Everything seems to be working fine now, will retest and report back.

**Comment by bloatware** [comment]
I'm about to commit, but need your opinion. Finally, the less effort way seems to be integrating assets as Forms of special user-definable types (`js`, `xml` etc). Few advantages over `txp_css` storage:

- that's how `rah_external_output` stores its snippets, so many users are used to it;
- more flexible: a `js`-type form could be used as inline script (via `<txp:output_form />`) as well as external asset;
- no new tab/pane/strings needed, just a new pref with some help text.

**Comment by Bloke** [comment]
Sounds good - how does that affect themes? Wasn't the point of this to allow themes to use external assets? I've lost the thread a bit so please correct me if I'm off-base.

**Comment by Bloke** [comment]
Sorry. Carry on. Just a momentary brain implosion.

**Comment by Bloke** [comment]
I'm fine with `<txp:asset />`. One caveat: are we going against the grain by referring to these things as assets or is it standard nomenclature in the CMS world? Digital Asset Management, for example, is media files like images and videos. In video editing, an asset is an imported media clip on the timeline. Don't want to get caught out calling custom Form snippets and CSS 'assets' if it's more conventional to refer to media with that name. Though, quite what we call them I'm not sure; 'output' seems too generic.

The decision on filename will be down to if we envisage using it for anything more than just serving these items. If not, `asset.php` (singular) is fine, subject to the above caveat.

**Comment by petecooper** [comment]
Any mileage in `component`?

**Comment by Bloke** [comment]
It has merit. Thesaurus.com suggests these synonyms for further consideration:

* ingredient
* item
* piece
* segment
* unit
* constituent

*shrug*

**Comment by petecooper** [comment]
Actually, `item` works, too.

(He says, being careful not to bike shed.)

**Comment by Bloke** [comment]
:) Yeah but the shed colour is actually important here so we don't bamboozle people.

**Comment by petecooper** [comment]
Well, we have Images for images, Files for other…erm…files, Forms for code chunks (which is a bonkers name, in my humble opinion but I've learned to love it), and as I understand it this issue is restricted to management of presentational files for themes (broadly). It's not a leap to say this can encompass front-side and admin-side themes, so I'm erring toward Asset(s) and `<txp:asset />` at the moment.

And if the `asset.php` filename trips up overly-sensitive content filters, then so be it.

**Comment by philwareham** [comment]
OK, so not too familiar with this patch, but I am currently building a [Bootstrap 4 framework as theme](https://github.com/philwareham/textpattern-bootstrap) for Textpattern and have an `assets` directory within the theme directory. That directory contains CSS, JS, images for example as subdirectories like so:

    themes
    → textpattern-bootstrap
        → assets
            → css
            → js
            → img
        → pages
        → styles
        ...etc...

**Comment by bloatware** [comment]
No problem expected with the directory structure, all new "assets" would be subdirectories of `forms`:
```
themes
→ textpattern-bootstrap
    → assets
        → css
        → js
        → img
    → forms
        → article
        ...etc...
        → js
        → xml
    → pages
    ...etc...
```

**Comment by NicolasGraph** [comment]
While `<txp:asset />` seems ok to me but  as a tag, just remember that in [the Skin package](https://github.com/textpattern/textpattern/tree/dev/textpattern/vendors/Textpattern/Skin) files [I used "asset(s)"](https://github.com/textpattern/textpattern/blob/dev/textpattern/vendors/Textpattern/Skin/Skin.php#L130) to talk about pages and/or forms and/or styles. Assets, here, are just first level theme "components".

**Comment by philwareham** [comment]
OK, so files only become visible and directly editable in the control panel *if* they are housed in the `forms` directory, yes?

Not sure about that, to be honest. These assets are a separate thing from Textpattern forms. I'd rather they were housed in their own silo.

Forgive me if I'm understanding incorrectly.

**Comment by Bloke** [comment]
Yeah, dammit I'd forgotten that the Skin package uses the term 'asset' to refer to all Pages/Forms/Styles. That's potentially messy/confusing for future code maintainability.

If we go with something like `component` or `item`, that's fine. But if we go with `asset` to refer to a custom Form type then, ummm... global search 'n replace anyone in the Skin package?!

**Comment by bloatware** [comment]
There is no new `Asset` class to define, so no name clashes to fear. We just need meaningful tag+file names to satisfy our users and search engines.

Internally, these `assets` are special-type forms stored in `txp_form` table and editable on Forms tab as any other form. If exported to disk, they will go to their type subdirectory (e.g. `js`) of `forms` directory, exactly as `article`, `misc` and other standard forms do. From there they can be served as flat files which is fine if their content is static. But they can also be parsed and generated dynamically when served from db, unlike the ordinary "assets".

Say, you want to create a (dynamic) sitemap. Currently, you need a specially crafted page and a section to do it, which is overkill. Or you somehow generate a `sitemap.xml` file and update it periodically. Now, you could just create a `xml`-type `sitemap` form (using `<txp:article_custom />` and other tags inside) and redirect SE to something like `asset.php?n=sitemap`.

**Comment by phiw13** [comment]
> Any mileage in component? Suitable vagueness?

I like the sound of that. `component` as a building brick.

(never ben a fan of `asset` or `assets`)

**Comment by bloatware** [comment]
Now we could image storing them in a new `assets` table/directory. It means more code/strings, but looks doable.

**Comment by bloatware** [comment]
Sitemaps are not building block :-/

**Comment by philwareham** [comment]
`component` as the naming works for me.

**Comment by Bloke** [comment]
> There is no new Asset class to define, so no name clashes to fear.

Sure, I get that. I was just concerned that an outsider/future developer viewing the code and seeing mention of an Asset class in a theme might misconstrue that as something to do with 'asset' as in the proposed tag and other features we're talking about here. So my notion of renaming was just to avoid confusion.

A `component` might be a misnomer as well then. Although we can't really envisage all uses for this feature, a suitable name for the PHP file and tag will certainly aid comprehension of what it can be used for. The sitemap example is a fantastic use case. I wonder if element works here?

`element.php?n=sitemap` It's an element of the site, and we're serving the sitemap.

It's also applicable to JavaScript import/export and XML files and a bunch of other use cases.

As far as tags go: `<txp:element type="js" name="my_javascript_lib.js" />` or something works too. Same for any type of file I can think of.

Random thought: If this is processed as part of the Form handling code, then are shortcuts available with the form name directly? e.g. `<txp::my_javascript_lib />`? Just wonderin'.

**Comment by bloatware** [comment]
Let's [test](https://github.com/textpattern/textpattern/commit/15f33979aedd01dffa2e8cdcc882469fa8887dc3) it. If a better forms/assets separation is desirable, we can do it in two ways:

- still store components as forms (both db and fs), but create a new "components" tab in admin interface.
- or introduce a new class with its own table/directory which could be called `assets`. Note, however, that any additional subdirectories (e.g. `img`) of `assets` would not be importable into db, hence not editable nor "duplicatable" via txp admin interface atm.

**Comment by bloatware** [comment]
Re `component.php` or other `.php` script: do we really need a new one? We could output "assets" via `index.php` as well, adding some URL parameter, say `n` or `form`:
```
index.php?form=myForm         // outputs the asset in default context
mysection/mytitle?form=myForm // outputs the asset in article context
```
In the second case, if `myForm` is of `json` type, the article will be output in `JSON` format.

**Comment by philwareham** [comment]
The less files the better in my opinion.

**Comment by Bloke** [comment]
I'm fine with using `index.php`. LIke Phil says, the fewer files the better and the less chance there is of a name clash anywhere.

**Comment by bloatware** [comment]
So it is `index.php?f=myform` now. I have no further plans, if `f` is ok as URL parameter, the issue can be closed.
